Makefile

```diff
@@ -89,36 +89,48 @@ ifeq ($(TARGET_ARCH),x64)
 	TARGET_ARCH_DEBIAN = amd64
 endif
 
-TEMPORARY_DIRECTORY = release/.tmp
```

Comment:
This is how it would look like:

```diff
diff --git a/Makefile b/Makefile
index 3649137..6998c78 100644
--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,12 @@
 # ---------------------------------------------------------------------
+# Build configuration
+# ---------------------------------------------------------------------
+
+BUILD_DIRECTORY = release
+BUILD_TEMPORARY_DIRECTORY = $(BUILD_DIRECTORY)/.tmp
+BUILD_OUT_DIRECTORY = $(BUILD_DIRECTORY)/out
+
+# ---------------------------------------------------------------------
 # Application configuration
 # ---------------------------------------------------------------------
 
@@ -96,20 +104,20 @@ APPLICATION_VERSION_DEBIAN = $(shell echo $(APPLICATION_VERSION) | tr "-" "~")
 # Rules
 # ---------------------------------------------------------------------
 
-release:
+$(BUILD_DIRECTORY):
 	mkdir $@
 
-release/.tmp: | release
+$(BUILD_TEMPORARY_DIRECTORY): | $(BUILD_DIRECTORY)
 	mkdir $@
 
-release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies: | release
+$(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies: | $(BUILD_DIRECTORY)
 	mkdir $@
 
-release/out: | release
+$(BUILD_OUT_DIRECTORY): | $(BUILD_DIRECTORY)
 	mkdir $@
 
-release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies/node_modules: package.json npm-shrinkwrap.json \
-	| release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies
+$(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies/node_modules: package.json npm-shrinkwrap.json \
+	| $(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies
 	./scripts/unix/dependencies-npm.sh -p \
 		-r "$(TARGET_ARCH)" \
 		-v "$(ELECTRON_VERSION)" \
@@ -117,33 +125,34 @@ release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies/node_modules: pa
 		-t electron \
 		-s "$(TARGET_PLATFORM)"
 
-release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies/bower_components: bower.json \
-	| release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies
+$(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies/bower_components: bower.json \
+	| $(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies
 	./scripts/unix/dependencies-bower.sh -p -x $(dir $@)
 
-release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-app: \
-	release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies/node_modules \
-	release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies/bower_components \
-	| release
+$(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-app: \
+	$(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies/node_modules \
+	$(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-dependencies/bower_components \
+	| $(BUILD_DIRECTORY)
 	./scripts/unix/electron-create-resources-app.sh -s . -f "$(APPLICATION_FILES)" -o $@
 	$(foreach prerequisite,$^,cp -rf $(prerequisite) $@;)
 
-release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-app.asar: \
-	release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-app \
-	| release
+$(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-app.asar: \
+	$(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-app \
+	| $(BUILD_DIRECTORY)
 	./scripts/unix/electron-create-asar.sh -d $< -o $@
 
-release/electron-$(ELECTRON_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).zip: | release
+$(BUILD_DIRECTORY)/electron-$(ELECTRON_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).zip: \
+	| $(BUILD_DIRECTORY)
 	./scripts/unix/electron-download-package.sh \
 		-r "$(TARGET_ARCH)" \
 		-v "$(ELECTRON_VERSION)" \
 		-s "$(TARGET_PLATFORM)" \
 		-o $@
 
-release/$(APPLICATION_NAME)-$(TARGET_PLATFORM)-$(TARGET_ARCH): \
-	release/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-app.asar \
-	release/electron-$(ELECTRON_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).zip \
-	| release
+$(BUILD_DIRECTORY)/$(APPLICATION_NAME)-$(TARGET_PLATFORM)-$(TARGET_ARCH): \
+	$(BUILD_DIRECTORY)/electron-$(TARGET_PLATFORM)-$(TARGET_ARCH)-app.asar \
+	$(BUILD_DIRECTORY)/electron-$(ELECTRON_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).zip \
+	| $(BUILD_DIRECTORY)
 ifeq ($(TARGET_PLATFORM),darwin)
 	./scripts/darwin/electron-configure-package-darwin.sh -p $(word 2,$^) -a $< \
 		-n "$(APPLICATION_NAME)" \
@@ -162,25 +171,25 @@ ifeq ($(TARGET_PLATFORM),linux)
 		-o $@
 endif
 
-release/$(APPLICATION_NAME)-$(TARGET_PLATFORM)-$(TARGET_ARCH)-rw.dmg: \
-	release/$(APPLICATION_NAME)-darwin-$(TARGET_ARCH) \
-	| release
+$(BUILD_DIRECTORY)/$(APPLICATION_NAME)-$(TARGET_PLATFORM)-$(TARGET_ARCH)-rw.dmg: \
+	$(BUILD_DIRECTORY)/$(APPLICATION_NAME)-darwin-$(TARGET_ARCH) \
+	| $(BUILD_DIRECTORY)
 	./scripts/darwin/electron-create-readwrite-dmg.sh -p $< -o $@ \
 		-n "$(APPLICATION_NAME)" \
 		-i assets/icon.icns \
 		-b assets/osx/installer.png
 
-release/out/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-darwin-$(TARGET_ARCH).zip: \
-	release/$(APPLICATION_NAME)-darwin-$(TARGET_ARCH) \
-	| release/out
+$(BUILD_OUT_DIRECTORY)/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-darwin-$(TARGET_ARCH).zip: \
+	$(BUILD_DIRECTORY)/$(APPLICATION_NAME)-darwin-$(TARGET_ARCH) \
+	| $(BUILD_OUT_DIRECTORY)
 ifdef CODE_SIGN_IDENTITY
 	./scripts/darwin/electron-sign-app.sh -a $</$(APPLICATION_NAME).app -i "$(CODE_SIGN_IDENTITY)"
 endif
 	./scripts/darwin/electron-installer-app-zip.sh -a $</$(APPLICATION_NAME).app -o $@
 
-release/out/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-darwin-$(TARGET_ARCH).dmg: \
-	release/$(APPLICATION_NAME)-$(TARGET_PLATFORM)-$(TARGET_ARCH)-rw.dmg \
-	| release/out
+$(BUILD_OUT_DIRECTORY)/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-darwin-$(TARGET_ARCH).dmg: \
+	$(BUILD_DIRECTORY)/$(APPLICATION_NAME)-$(TARGET_PLATFORM)-$(TARGET_ARCH)-rw.dmg \
+	| $(BUILD_OUT_DIRECTORY)
 ifdef CODE_SIGN_IDENTITY
 	./scripts/darwin/electron-sign-dmg.sh \
 		-n "$(APPLICATION_NAME)" \
@@ -189,19 +198,19 @@ ifdef CODE_SIGN_IDENTITY
 endif
 	./scripts/darwin/electron-create-readonly-dmg.sh -d $< -o $@
 
-release/out/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-linux-$(TARGET_ARCH).zip: \
-	release/$(APPLICATION_NAME)-linux-$(TARGET_ARCH) \
-	| release/out release/.tmp
-	TMPDIR=release/.tmp ./scripts/linux/electron-installer-appimage.sh -p $< -o $@ \
+$(BUILD_OUT_DIRECTORY)/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-linux-$(TARGET_ARCH).zip: \
+	$(BUILD_DIRECTORY)/$(APPLICATION_NAME)-linux-$(TARGET_ARCH) \
+	| $(BUILD_OUT_DIRECTORY) $(BUILD_TEMPORARY_DIRECTORY)
+	TMPDIR=$(BUILD_TEMPORARY_DIRECTORY) ./scripts/linux/electron-installer-appimage.sh -p $< -o $@ \
 		-n "$(APPLICATION_NAME)" \
 		-d "$(APPLICATION_DESCRIPTION)" \
 		-r "$(TARGET_ARCH)" \
 		-b "$(APPLICATION_NAME_LOWERCASE)" \
 		-i assets/icon.png
 
-release/out/$(APPLICATION_NAME_LOWERCASE)-electron_$(APPLICATION_VERSION_DEBIAN)_$(TARGET_ARCH_DEBIAN).deb: \
-	release/$(APPLICATION_NAME)-linux-$(TARGET_ARCH) \
-	| release/out
+$(BUILD_OUT_DIRECTORY)/$(APPLICATION_NAME_LOWERCASE)-electron_$(APPLICATION_VERSION_DEBIAN)_$(TARGET_ARCH_DEBIAN).deb: \
+	$(BUILD_DIRECTORY)/$(APPLICATION_NAME)-linux-$(TARGET_ARCH) \
+	| $(BUILD_OUT_DIRECTORY)
 	./scripts/linux/electron-installer-debian.sh -p $< -r "$(TARGET_ARCH)" \
 		-c scripts/build/debian/config.json \
 		-o $(dir $@)
@@ -234,12 +243,12 @@ electron-develop:
 	./scripts/unix/dependencies-bower.sh
 
 ifeq ($(TARGET_PLATFORM),darwin)
-electron-installer-app-zip: release/out/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).zip
-electron-installer-dmg: release/out/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).dmg
+electron-installer-app-zip: $(BUILD_OUT_DIRECTORY)/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).zip
+electron-installer-dmg: $(BUILD_OUT_DIRECTORY)/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).dmg
 endif
 ifeq ($(TARGET_PLATFORM),linux)
-electron-installer-appimage: release/out/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).zip
-electron-installer-debian: release/out/$(APPLICATION_NAME_LOWERCASE)-electron_$(APPLICATION_VERSION_DEBIAN)_$(TARGET_ARCH_DEBIAN).deb
+electron-installer-appimage: $(BUILD_OUT_DIRECTORY)/$(APPLICATION_NAME)-$(APPLICATION_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).zip
+electron-installer-debian: $(BUILD_OUT_DIRECTORY)/$(APPLICATION_NAME_LOWERCASE)-electron_$(APPLICATION_VERSION_DEBIAN)_$(TARGET_ARCH_DEBIAN).deb
 endif
 
 info:
@@ -249,4 +258,4 @@ info:
 	@echo "Target arch     : $(TARGET_ARCH)"
 
 clean:
-	rm -rf release
+	rm -rf $(BUILD_DIRECTORY)
```

I see a couple of flaws already:

- The directory creation rules hardcode a certain relationship. For example:

```make
$(BUILD_TEMPORARY_DIRECTORY): | $(BUILD_DIRECTORY)
	mkdir $@
```

This implies that `$(BUILD_TEMPORARY_DIRECTORY)` is inside `$(BUILD_DIRECTORY)`, but of course the user could configure this differently, in which case the rule stops making sense. This could be fixed by not using any `|` pre-requisites on the build directories.

- A certain "build directory" might contain any level of subdirectories. For example, I could declare that the "out directory" is `$(BUILD_DIRECTORY)/foo/bar/baz/out`, in which case our rules would fail because they don't include `mkdir -p`.

What do you think? I like how this reduces duplication, but I wonder if its the right thing to do.